server/portal/apps/projects/workspace_operations/shared_workspace_operations.py

```diff
@@ -0,0 +1,329 @@
+from portal.utils.encryption import createKeyPair
+from portal.libs.agave.utils import service_account
+from tapipy.tapis import Tapis
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from portal.apps.onboarding.steps.system_access_v3 import create_system_credentials, register_public_key
+
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+def set_workspace_permissions(client: Tapis, username: str, system_id: str, role: str):
+    """Apply read/write/execute permissions to a user on a system."""
+
+    system_pems = {
+         "reader": ["READ", "EXECUTE"],
+         "writer": ["READ", "EXECUTE"]
+    }
+
+    files_pems = {
+         "reader": "READ",
+         "writer": "MODIFY"
+    }
+
+    logger.info(f"Adding {username} permissions to Tapis system {system_id}")
+    client.systems.grantUserPerms(
+        systemId=system_id,
+        userName=username,
+        permissions=system_pems[role])
+
+    if role == "reader":
+        client.systems.revokeUserPerms(systemId=system_id,
+                                       userName=username,
+                                       permissions=["MODIFY"])
+        client.files.deletePermissions(systemId=system_id,
+                                       path="/",
+                                       username=username)
+
+    client.files.grantPermissions(
+        systemId=system_id,
+        path="/",
+        username=username,
+        permission=files_pems[role]
+    )
+
+
+def set_workspace_acls(client, system_id, path, username, operation, role):
+
+    operation_map = {
+         "add": "ADD",
+         "remove": "REMOVE"
+    }
+
+    acl_string_map = {
+         "reader": f"d:u:{username}:rX,u:{username}:rX",
+         "writer": f"d:u:{username}:rwX,u:{username}:rwX",
+         "none": f"d:u:{username},u:{username}"
+    }
+
+    client.setFacl(systemId=system_id,
+                   path=path,
+                   operation=operation_map[operation],
+                   recursionMethod="PHYSICAL",
```

Comment:
> recursionMethod="PHYSICAL",

This doesn't appear to be applying recursively. Is there another method?

Reply:
Here you can see I applied the "GUEST" role to `sal_test`, but it only changed permissions to `r-x` on the parent project folder
```
[sal@cloud CEPV3-DEV-12]$ ll
total 4
-rw-rw----+ 1 sal TACC-ACI  532 Mar 28 11:54  chimeraxapp-0-0-1_2023-02-28t17-15-15-dcvserver-559dee05-74f1-457a-beb2-509f7618319a-007.out
drwxrws--x+ 2 sal TACC-ACI 4096 Mar 28 12:08  test
drwxrws--x+ 2 sal TACC-ACI 4096 Mar 28 12:11  test2
drwxrws--x+ 2 sal TACC-ACI 4096 Mar 28 11:54 'Test fol'
[sal@cloud CEPV3-DEV-12]$ ll
total 4
-rw-rw----+ 1 sal TACC-ACI  532 Mar 28 11:54  chimeraxapp-0-0-1_2023-02-28t17-15-15-dcvserver-559dee05-74f1-457a-beb2-509f7618319a-007.out
drwxrws--x+ 2 sal TACC-ACI 4096 Mar 28 12:08  test
drwxrws--x+ 2 sal TACC-ACI 4096 Mar 28 12:11  test2
drwxrws--x+ 2 sal TACC-ACI 4096 Mar 28 11:54 'Test fol'
[sal@cloud CEPV3-DEV-12]$ getfacl ../CEPV3-DEV-12
# file: ../CEPV3-DEV-12
# owner: wma_prtl
# group: TACC-ACI
# flags: -s-
user::rwx
user:wma_prtl:rwx
user:sal:rwx
user:sal_test:r-x
group::rwx
mask::rwx
other::--x
default:user::rwx
default:user:wma_prtl:rwx
default:user:sal:rwx
default:user:sal_test:r-x
default:group::rwx
default:mask::rwx
default:other::--x

[sal@cloud CEPV3-DEV-12]$ getfacl test
# file: test
# owner: sal
# group: TACC-ACI
# flags: -s-
user::rwx
user:wma_prtl:rwx
user:sal:rwx
group::rwx
mask::rwx
other::--x
default:user::rwx
default:user:wma_prtl:rwx
default:user:sal:rwx
default:group::rwx
default:mask::rwx
default:other::--x
```